src/header/parameters.rs

```diff
@@ -33,12 +35,98 @@ pub(crate) struct Parameters<T> {
     pub(crate) x509_certificate_sha256_thumbprint: Option<HeaderValue<[u8; 32]>>,
     /// `typ` parameter defined in section 4.1.9 of JWS and section 4.1.11 of
     /// JWE
-    pub(crate) typ: Option<HeaderValue<MediaTypeBuf>>,
+    pub(crate) typ: Option<HeaderValue<MediaTypeWithMaybeStrippedApplicationTopLevel>>,
     /// `cty` parameter defined in section 4.1.10 of JWS and section 4.1.12 of
     /// JWE
-    pub(crate) content_type: Option<HeaderValue<MediaTypeBuf>>,
+    pub(crate) content_type: Option<HeaderValue<MediaTypeWithMaybeStrippedApplicationTopLevel>>,
     // additional parameters specific to JWS or JWE (e.g. `enc` in JWE)
     pub(crate) specific: T,
     // an untyped list of other values that are not understood by this implementation
     pub(crate) additional: BTreeMap<String, HeaderValue<Value>>,
 }
+
+/// A wrapper type that incorporates the special handling of `application/X`
+/// media types in `typ` and `cty` header parameters as defined in [Section
+/// 4.1.9 of RFC 7515][1].
+///
+/// See [#128].
+///
+/// If the mediatype starts with `application/` and no other `/` is present,
+/// implementations should strip the `application/` to save space.
+///
+/// Since this is a serialization detail, we abstract it away, because the
+/// meaning remains the same and this way users can easily use the mediatype
+/// crate.
+///
+/// [#128]: <https://github.com/minkan-chat/jose/issues/128>
+/// [1]: <https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.9>
+#[derive(Debug)]
+pub(crate) struct MediaTypeWithMaybeStrippedApplicationTopLevel(pub(crate) MediaTypeBuf);
+
+impl<'de> Deserialize<'de> for MediaTypeWithMaybeStrippedApplicationTopLevel {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let raw: Cow<'_, str> = Cow::deserialize(deserializer)?;
+        let corrected = if raw.find("/").is_none() {
+            alloc::format!("application/{raw}")
+        } else {
+            raw.to_string()
+        };
+        let inner = MediaTypeBuf::from_string(corrected).map_err(D::Error::custom)?;
+        Ok(Self(inner))
+    }
+}
+
+impl Serialize for MediaTypeWithMaybeStrippedApplicationTopLevel {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let correct = self.0.to_string();
+        if self.0.ty() == mediatype::names::APPLICATION
+            && correct.chars().filter(|c| *c == '/').count() == 1
+        {
+            let raw = correct.split_once('/').expect("contains one slash").1;
```

Comment:
Can't you use the `MediaTypeBuf.subty` field so you don't have to do the parsing by yourself?